ghcide/src/Development/IDE/Plugin/Completions.hs

```diff
@@ -131,35 +132,49 @@ getCompletionsLSP ide plId
     fmap Right $ case (contents, uriToFilePath' uri) of
       (Just cnts, Just path) -> do
         let npath = toNormalizedFilePath' path
-        (ideOpts, compls) <- liftIO $ runIdeAction "Completion" (shakeExtras ide) $ do
+        (ideOpts, compls, moduleExports) <- liftIO $ runIdeAction "Completion" (shakeExtras ide) $ do
             opts <- liftIO $ getIdeOptionsIO $ shakeExtras ide
             localCompls <- useWithStaleFast LocalCompletions npath
             nonLocalCompls <- useWithStaleFast NonLocalCompletions npath
             pm <- useWithStaleFast GetParsedModule npath
             binds <- fromMaybe (mempty, zeroMapping) <$> useWithStaleFast GetBindings npath
             exportsMapIO <- fmap(envPackageExports . fst) <$> useWithStaleFast GhcSession npath
             exportsMap <- mapM liftIO exportsMapIO
+            let moduleExports = buildModouleExportMap exportsMap
             let exportsCompItems = foldMap (map (fromIdentInfo uri) . Set.toList) . Map.elems . getExportsMap <$> exportsMap
                 exportsCompls = mempty{anyQualCompls = fromMaybe [] exportsCompItems}
             let compls = (fst <$> localCompls) <> (fst <$> nonLocalCompls) <> Just exportsCompls
-            pure (opts, fmap (,pm,binds) compls)
-        case compls of
-          Just (cci', parsedMod, bindMap) -> do
+            pure (opts, fmap (,pm,binds) compls, moduleExports)
+        case (compls, moduleExports) of
+          (Just (cci', parsedMod, bindMap), mExports) -> do
             pfix <- VFS.getCompletionPrefix position cnts
             case (pfix, completionContext) of
               (Just (VFS.PosPrefixInfo _ "" _ _), Just CompletionContext { _triggerCharacter = Just "."})
                 -> return (InL $ List [])
               (Just pfix', _) -> do
                 let clientCaps = clientCapabilities $ shakeExtras ide
                 config <- getCompletionsConfig plId
-                allCompletions <- liftIO $ getCompletions plId ideOpts cci' parsedMod bindMap pfix' clientCaps config
+                allCompletions <- liftIO $ getCompletions plId ideOpts cci' parsedMod bindMap pfix' clientCaps config mExports
                 pure $ InL (List allCompletions)
               _ -> return (InL $ List [])
           _ -> return (InL $ List [])
       _ -> return (InL $ List [])
 
 ----------------------------------------------------------------------------------------------------
 
+identInfoToKeyVal :: IdentInfo -> (T.Text, T.Text)
+identInfoToKeyVal IdentInfo {rendered, moduleNameText} =
+  (moduleNameText, rendered)
+
+buildModouleExportMap:: Maybe (ExportsMap) -> DM.Map T.Text [T.Text]
+buildModouleExportMap (Just exportsMap) = do
```

Comment:
What does this function do?

Reply:
I had a typo in the name, however, it builds a map of `"module.name" -> ["functions" .. ]` which I use to look up the available functions for that module

Reply:
And why is it needed? Can't you use the ExportsMap as it is?

Reply:
I believe I need it because I have to group on `"module.name"` and use that as the key to look up. `ExportsMap` as is seems to be a map of `"function-name" -> [{moduleNameText: "module-name",  name: "function-name"}]`

Reply:
Ah yes, you are right. 

But why build a map to do only one lookup and then throw it away? 

* Cost of building the map: O(NlogN)
* Cost of doing the lookup by scanning the original exports map: O(N)

Neither is good enough, you want O(logN). How? Probably by changing the representation of the exports map to add a second index, the module name.

Reply:
I don't think I follow your suggestion. Do you mean to add the moduleName as keys, or another layer of indexes?

Reply:
Add another field to `ExportsMap` containing a map from modules to exports. 

But a better solution would be to avoid using the exports map at all. Instead, retrieve the `ModIface` for the module and extract the list of exports. To do that, you first need to map the module name to a path using `GetLocatedImports`. At this point, I would consider moving the code to a new plugin.

Reply:
Do you think the first suggestion would be a good first pass? With potential follow up commits to pull out a plugin?

Reply:
The ExportsMap only indexes package modules. You will need to fetch the ModIface for local modules. So both changes are needed

Reply:
I took a first pass at this. There is still more work to do. As of this comment, I am just generating the map based on the already existing map. I also see that this does not work for local modules, as you suggested.

Reply:
Second pass - generate the map from the modIface rather than the existing map.. This still does not account for local modules

Reply:
(sorry for all the of the noise) - third pass - enable local module imports

Reply:
OK, I now generate the map for local modules as well. I will take a look at your other comments.